### ISSUE TITLE
Support no_std where possible (specifically for wasm32v1-none)

### ISSUE THREAD

**Issue by Dinnerbone** [gfx-rs/wgpu#6826]
**Is your feature request related to a problem? Please describe.**
`naga`, `wgpu` and related crates should have `std` as a (default) feature, specifically for `wasm32v1-none` support.

**Describe the solution you'd like**
There's a hope to get the wasm ecosystem slowly shifting from `wasm32-unknown-unknown` to [`wasm32v1-none`](https://doc.rust-lang.org/nightly/rustc/platform-support/wasm32v1-none.html), which is hopefully stabilizing in rust 1.84. This is primarily because `wasm32-unknown-unknown` is [something of a moving target and can change the browser requirements at will](https://blog.rust-lang.org/2024/09/24/webassembly-targets-change-in-default-target-features.html), which we recently discovered when older Safari stopped working 😅

The main differences between `wasm32v1-none` and `wasm32-unknown-unknown` are:
- It does not enable any wasm features by default (which means it won't arbitrarily break browser compatibility with rust updates)
- It has no std (as opposed to a stubbed pseudo-std)

Unfortunately, no_std is both a blessing and a curse: it's neat to get things using core/alloc where possible, but it sucks that the majority of the ecosystem needs churn to make this viable.

I've done a quick investigation and I think it's plausible to get at least web working in `no_std`, but potentially other platforms too. `wgpu-types` is the easiest to get `no_std`, it only needs it for one specific dx12 config. `naga` is possibly the second easiest crate, but it requires some work in dependencies. wgpu-core and wgpu-hal will need more work mostly surrounding the use of mutexes.

**Describe alternatives you've considered**
We could just _not_, but that means it's impossible to allow users of wgpu to have any control over which browsers they support.
We can also limit this to a specific set of features, there's no real need today for DX12 and `no_std`, for example.

**Comment by bushrat011899** [comment]
I think there's value in having even just a subset of `wgpu` be `no_std` compatible. For context, I'm working on making [Bevy `no_std` compatible](https://github.com/bevyengine/bevy/issues/15460), and a big line in the sand right now is anything related to our rendering crates, since they're built around `wgpu` and would be almost impossible to make independent of `wgpu` for the sake of an `std` feature.

**Comment by rdrpenguin04** [comment]
This would also be useful for porting to game consoles, particularly console homebrew.

**Comment by cwfitzgerald** [comment]
We're definitely interested in having this, but will need to be from an outside contributor, so PR's welcome!

I think that it is important that when the first PR comes in to partially support no-std, we have a CI job to make sure we don't break it in the future. That could be as simple as adding this wasm target to CI.

**Comment by brody2consult** [comment]
@bushrat011899 wrote in https://github.com/gfx-rs/wgpu/pull/6892#issuecomment-2585593107:
> Thanks for the assistance in getting it over the line! I'm planning on looking into `no_std` for `naga` next, but I suspect that'll be a much longer process.

@bushrat011899 I am working on this update for `wgpu-core`, hope to have a PR ready with this soon. I am taking a slightly different approach and would appreciate your feedback on this. I have also raised a couple more PRs to help prepare for these updates (see recent references above).

I may consider hacking on `wgpu-hal` as well, if you are not already working on this.

Hope I am not stepping on any toes.

**Comment by bushrat011899** [comment]
> Hope I am not stepping on any toes.

Not at all! Just happy to have more people working on `no_std` support!

**Comment by brody2consult** [comment]
FYI I have now got no-std build working with naga in my workarea, with really ugly hacking - unfortunately I need to take rest after a very long day, hope to raise a PR for this before the weekend. I have ended up learning quite a bit from this :)

---

P.S. Here is the branch I pushed to backup my workarea: https://github.com/brodycj/wgpu/tree/wipsave34asdf

This is based on PR #6926 for wgpu-core, only recommended for the super-brave!

**Comment by kpreid** [comment]
I’d like to recommend a change to the problem statement here:

> `naga`, `wgpu` and related crates should have `std` as a (default) feature,

While having a feature named `std` makes sense for some cases, like libraries which provide a trait and implement it for `std` types, I think that it is not a good pattern in general, because it is not clear when dependents should enable them. Instead, as much as possible, libraries should have features named for the *actual functionality they enable*, and document which of those features require `std`.

**Comment by cwfitzgerald** [comment]
Seems reasonable to me - we also have things like backends which can't be `no-std` (dx12 will always have std, for example)

**Comment by bushrat011899** [comment]
> Instead, as much as possible, libraries should have features named for the _actual functionality they enable_, and document which of those features require `std`.

Agreed, but I would definitely recommend doing this from the `Cargo.toml` through features. Using the DX12 backend as an example:

```toml
[features]
std = ["foo"]
dx12 = ["std", "bar"]
```

This clearly communicates that the `dx12` feature requires `std`, and also keeps the `lib.rs` file clean:

```rust
#![no_std]

#[cfg(feature = "std")]
extern crate std;
```

This does mean users can _see_ the `std` feature and have the option of enabling it even when it might not give them functionality, but I think that's a reasonable compromise. Like any crate feature though, libraries should activate the bare minimum they need, and gate functionality where reasonable to avoid activating too many features.

**Comment by kpreid** [comment]
>  keeps the lib.rs file clean

`wgpu` already solves much more complex `cfg` problems by using the build script to create cfg aliases that do not appear in the `[features]` table.

> This clearly communicates that the `dx12` feature requires `std`

I agree that this is useful documentation to have, but I do not think a feature should be added solely in order to achieve it.

**Comment by bushrat011899** [comment]
Perfectly reasonable opinion, I do appreciate reducing the quantity of features. I recently merged a change to Bevy which completely hides the `portable-atomic` feature, relying exclusively on `cfg(has_atomic_feature = "...")` to bring in (and use) the dependency only when it's needed.

**Comment by ErichDonGubler** [comment]
Some progress was claimed to be made in #7232, which was subsumed by <https://github.com/gfx-rs/wgpu/pull/7043>. I'm not sure if this is done yet, so calling out my immediate inability to determine what scope is left here.

**Comment by kpreid** [comment]
There is definitely work remaining to be done in order to meet the goal of running wgpu, with a web backend, on `wasm32v1-none`. It has many dependencies that are not `no_std`. For example, here is are some of the errors from a simple `cargo check`:

```
$ cargo check -p wgpu --no-default-features --target=wasm32v1-none
error[E0463]: can't find crate for `std`
  --> /Users/kpreid/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/raw-window-handle-0.6.2/src/lib.rs:34:1
   |
34 | extern crate std;
   | ^^^^^^^^^^^^^^^^^ can't find crate
   |
   = note: the `wasm32v1-none` target may not support the standard library

error[E0463]: can't find crate for `std`
  |
  = note: the `wasm32v1-none` target may not support the standard library
  = note: `std` is required by `parking_lot_core` because it does not declare `#![no_std]`

error[E0463]: can't find crate for `std`
   --> /Users/kpreid/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/thiserror-2.0.11/src/lib.rs:275:1
    |
275 | extern crate std;
    | ^^^^^^^^^^^^^^^^^ can't find crate
    |
    = note: the `wasm32v1-none` target may not support the standard library
```

(Note that because these errors are in different crates, any single build command may abort early and not print the rest of them.)

This means that you cannot even use the WebGPU backend. If we also want to support WebGL, then additional work is needed:

* `naga` needs `no_std` support, including removing or PRing std dependencies like `termcolor`
* The `gles` backend needs to be modified to, when working with `webgl`, avoid using `Mutex` and use `RefCell` instead, and other such changes.

**Comment by bushrat011899** [comment]
Opened a [PR](https://github.com/brendanzab/codespan/pull/374) to add `no_std` support to `codespan-reporting`, which would greatly simplify `no_std` in Naga itself. Noting it here so people interested in this issue are able to review the changes there if they wish.

**Comment by bushrat011899** [comment]
The `no_std` PR I opened for `codespan-reporting` has been successfully [merged](https://github.com/brendanzab/codespan/pull/374), with a plan to release a new version in a week or so. Once that's done, `no_std` support for `naga` can make further progress.

**Comment by bushrat011899** [comment]
Working on a `no_std` [PR for `arbitrary`](https://github.com/rust-fuzz/arbitrary/pull/216) so that it isn't a blocker for some of the `naga` features.

**Comment by bushrat011899** [comment]
Have a `no_std` compatible implementation of `round_ties_even` in a PR for `num-traits`. If that PR is accepted that's the final blocker for `no_std` support in Naga. If not, the content of the PR could be easily used within Naga, but it would be nicer to solve this upstream.

**Comment by bushrat011899** [comment]
All dependencies in Naga now have `no_std` PRs either opened or merged awaiting release. The SPIR-V features are the only ones with reliance on `std` once the above PRs are merged, due to `DebugInfo::file_name`, `Options::block_ctx_dump_prefix`, and the writing of block context dumps (I'm unfamiliar with SPIR-V so I don't understand whether that's required functionality).

**Comment by bushrat011899** [comment]
So, the two PRs I have open against this issue I think resolve it. I've published a demo [here](https://github.com/bushrat011899/wasm32v1-none-wgpu-example) which shows running the Hello Triangle example with my branches of Wgpu and Winit (which I've also submitted a `no_std` PR for). A live demo of that example is available through GitHub pages [here](https://bushrat011899.github.io/wasm32v1-none-wgpu-example/).

EDIT: I've updated the above example with a larger port of `wgpu_examples` to `no_std`. Below is a screenshot of the `water` example running on Firefox compiled for `wasm32v1-none`. All other examples shown run in Chromium based browsers, but certain ones fail on Firefox, possibly due to my Winit port or other browser incompatibilities.

<img width="2067" height="1203" alt="Image" src="https://github.com/user-attachments/assets/04e61909-979d-49c9-9a5c-8f59c459e02b" />
